src/shims/unix/foreign_items.rs

```diff
@@ -78,7 +78,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             "chdir" => {
                 let [path] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
                 let result = this.chdir(path)?;
-                this.write_scalar(result, dest)?;
+                this.write_io_result(result, dest)?;
```

Comment:
Is it really such a big win to move the io::Result handling out of the shim?

A priori I think I'd prefer the version that returns a Scalar, as that matches the C function we're implementing. We can have more helpers *inside* those functions that make them easier to write and avoid hard-coding `-1` anywhere, that seems like a more promising approach to me.

Reply:
I initially wanted to add such helpers, but then it got annoying for handling `i64`, `i32`, and `i64 but is actually target usize`, as that would require explicit type annotations and reliance on runtime checks (Scalar size vs destination size). I then thought it would be nicer to keep some semblence of result handling around, as now all the return types must match up again, like before moving to scalars. I'll finish this PR to its conclusion and open a separate PR for just using helpers everywhere so you can compare.